# Objet : Analyse des relations entre champs des données IRVE

## Contexte
- clarification du rôle des modèles de données dans les jeux de données (cf mise à jour récente des [guides data.gouv](https://guides.etalab.gouv.fr/qualite/documenter-les-donnees/))
- intégration d'une propriété "relationship" dans les schémas de données ([issue TableSchema](https://github.com/frictionlessdata/specs/issues/803) en cours de validation)
- création d'outils de contrôle des relations entre champs des jeux de données tabulaires (cf usage ci-dessous)

## Objectifs
- valider sur un cas réel l'utilisation d'un modèle de données en complément d'un schéma de données
- identifier les apports que pourraient avoir les contrôles de validation des relations entre champs

## Résultats
- la formalisation d'un modèle de données facilite la compréhension des données et des relations entre champs
- l'outil de contrôle permet d'améliorer significativement la qualité des données par la détection d'incohérences de relations
- l'analyse a posteriori permet de (re)trouver la logique de dépendance entre les colonnes qui minimise les incohérences
- les incohérences détectées sur le jeu de données IRVE restent faibles (inférieures à 2% des point de charge documentés - voir chapitre 4)

## Suite à donner
- Mettre à jour, valider et publier le modèle de données IRVE
- Définir les contrôles supplémentaires à intégrer pour toutes nouvelles données ainsi que pour le jeu complet
- Mettre en oeuvre les outils de contrôle

## Evolutions possibles 
- Ajouter dans les guides d'Etalab un guide pour les modèles de données 
- Intégrer dans les schémas de données la propriété "relationship" en cours de validation,
- Définir un indicateur qui mesure l'écart (existant / attendu) des relations entre champs

## Sommaire
*(liens actifs sur jupyter Notebook ou Nbviewer)*
- [1 - modèle de données](#1---modèle-de-données)
- [2 - Initialisation](#2---Initialisation)
- [3 - Séparation des pdc itinerance et hors itinerance](#3---Séparation-des-pdc-itinerance-et-hors-itinerance)
- [4 - Bilan initial intégrité](#4---Bilan-initial-intégrité)
- [5 - Séparation doublons pdc - date de maj](#5---Séparation-doublons-pdc---date-de-maj)
- [6 - Cohérence date maj stations](#6---Cohérence-date-maj-stations)
- [7 - Synthèse](#7---Synthèse)
- [8 - Exemples d erreurs résiduelles](#8---Exemples-d-erreurs-résiduelles)
- [Annexe : Initialisation du modèle de données 'mermaid'](#Annexe-:-Initialisation-du-modèle-de-données-'mermaid')

Ce Notebook peut être consulté sur [nbviewer](http://nbviewer.org/github/loco-philippe/Environmental-Sensing/tree/main/python/Validation/irve)

données utilisées : https://www.data.gouv.fr/fr/datasets/fichier-consolide-des-bornes-de-recharge-pour-vehicules-electriques/    
fichier : "*consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv*"

-----------------------
# 1 - modèle de données

Le modèle de données proposé ci-dessous est construit sur la base du schéma de données mis à disposition et du contenu du jeu de données.    
Il est à consolider en fonction de l'expertise des concepteurs et réutilisateurs (voir [guide méthodologique](https://github.com/loco-philippe/Environmental-Sensing/blob/main/property_relationship/FR_methodology.ipynb)).

*Notation:*
- *M : Mandatory - documentation obligatoire*
- *PK : Primary Key - identifiant unique de l'entité*
- *Root : champ fictif associé à une ligne du tableau*

In [3]:
# la variable 'irve' est définie en fin de Notebook
mermaid(irve)

In [217]:
def analyse_integrite(data):
    analyse = Analysis(data)
    dic_res = analyse.check_relationship(fields)
    data['ok'] = True
    for name, lis in dic_res.items():
        data[name] = True
        data.loc[lis, name] = False
        data['ok'] = data['ok'] & data[name]
        print('{:<50} {:>5}'.format(name, len(data) - data[name].sum()))
    return dic_res

------
# 2 - Initialisation
## initialisation des données
- lecture du fichier issu de l'api

In [311]:
from collections import Counter
from observation import Ilist, Iindex, Analysis
import pandas as pd
res = {}
chemin = 'https://raw.githubusercontent.com/loco-philippe/Environmental-Sensing/main/python/Validation/irve/'
res['date'] = '2023-03-03'
file = 'consolidation-etalab-schema-irve-statique-v-2.2.0-20230303.csv'
irve = pd.read_csv(chemin + file, sep=',', low_memory=False)
res['len_irve'] = len(irve)
print('nombre de lignes : ', res['len_irve']) 

nombre de lignes :  51276


## schéma de données
Le schéma de données restreint à la propriété 'relationship' et construit à partir du modèle de données est le suivants :

In [281]:
fields = [
 
 # relation unicité des pdl
 { "name": "index",
   "relationship" : { "parent" : "id_pdc_itinerance", "link" : "coupled" }},   
 # relations inter entités
 { "name": "contact_operateur",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nom_enseigne",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "coordonneesXY",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "id_station_itinerance",
   "relationship" : { "parent" : "id_pdc_itinerance",     "link" : "derived" }},
 # relations intra entité - station
 { "name": "nom_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "implantation_station",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "nbre_pdc",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "condition_acces",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "horaires",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 { "name": "station_deux_roues",
   "relationship" : { "parent" : "id_station_itinerance", "link" : "derived" }},
 # relations intra entité - localisation
 { "name": "adresse_station",
   "relationship" : { "parent" : "coordonneesXY",         "link" : "derived" }} ]

In [224]:
relations = ['index', 'contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 
             'nom_station', 'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 
             'id_pdc_itinerance', 'date_maj', 'last_modified']
champs =  ['contact_operateur', 'nom_enseigne', 'coordonneesXY', 'adresse_station', 'id_station_itinerance', 'nom_station',
           'implantation_station', 'nbre_pdc', 'condition_acces', 'horaires', 'station_deux_roues', 'id_pdc_itinerance', 
           'puissance_nominale', 'prise_type_ef', 'prise_type_2', 'prise_type_combo_ccs', 'prise_type_chademo', 
           'prise_type_autre', 'paiement_acte', 'paiement_autre', 'reservation',  'accessibilite_pmr', 'restriction_gabarit', 
           'date_maj', 'last_modified']

-----------------------
## 3 - Séparation des pdc itinerance et hors itinerance
- un peu moins de 1 % des points de charge sont hors itinerance

In [351]:
data = irve.loc[:, champs]
data[['id_station_itinerance','id_pdc_itinerance']] = data[['id_station_itinerance','id_pdc_itinerance']].astype('string')
#data[['last_modified','date_maj']] = data[['last_modified','date_maj']].astype('datetime64')
data['non_concerne'] = data['id_station_itinerance'].str.contains('oncern') | data['id_pdc_itinerance'].str.contains('oncern')

non_concerne = data[data['non_concerne']].loc[:, champs].reset_index()['index']
itinerance = data[~data['non_concerne']].loc[:, champs].reset_index()
itinerance = itinerance.loc[:, relations]
print('nombre de pdc hors itinerance : ', len(non_concerne))
print('nombre de pdc en itinerance   : ', len(itinerance))

nombre de pdc hors itinerance :  607
nombre de pdc en itinerance   :  50669


-----------------------
## 4 - Bilan initial intégrité
- 36 % des lignes sont erronées ( 18 116 )

In [314]:
res = analyse_integrite(itinerance)
itinerance_ok_1 = itinerance.loc[itinerance.ok, relations].reset_index(drop=True)
itinerance_1 = itinerance.loc[~itinerance.ok, relations].reset_index(drop=True)
itinerance = itinerance.loc[:, relations]
print("\nnombre d'enregistrements sans erreurs : ", len(itinerance_ok_1))
print("nombre d'enregistrements avec au moins une erreur : ", len(itinerance_1))
print("taux d'erreur : ", round(len(itinerance_1) / len(itinerance) * 100), ' %')

index - id_pdc_itinerance                          16123
contact_operateur - id_station_itinerance          10719
nom_enseigne - id_station_itinerance                7514
coordonneesXY - id_station_itinerance              11825
id_station_itinerance - id_pdc_itinerance            578
nom_station - id_station_itinerance                 1865
implantation_station - id_station_itinerance        1245
nbre_pdc - id_station_itinerance                    1458
condition_acces - id_station_itinerance               35
horaires - id_station_itinerance                    9869
station_deux_roues - id_station_itinerance         10968
adresse_station - coordonneesXY                     1360

nombre d'enregistrements sans erreurs :  32553
nombre d'enregistrements avec au moins une erreur :  18116
taux d'erreur :  36  %


-----------------------
## 5 - Séparation doublons pdc - date de maj
- la moitié des pdc en erreur sont liées aux doublons de pdc
- la suppression des doublons permet de diviser par 5 le nombre de lignes erronnées ( 2 166 )

In [336]:
itinerance_1['doublons_pdc'] = itinerance_1.sort_values(by='date_maj').duplicated('id_pdc_itinerance', keep='last')

doublons_pdc = itinerance_1[itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)['index']
itinerance_2 = itinerance_1[~itinerance_1['doublons_pdc']].loc[:, relations].reset_index(drop=True)
itinerance_1 = itinerance_1.loc[:, relations]
print('nombre de doublons pdc : ', len(doublons_pdc))
print('nombre de pdc sans doublon   : ', len(itinerance_2))

nombre de doublons pdc :  8301
nombre de pdc sans doublon   :  9815


In [337]:
res = analyse_integrite(itinerance_2)
itinerance_ok_3 = itinerance_2.loc[itinerance_2.ok, relations].reset_index(drop=True)
itinerance_3 = itinerance_2.loc[~itinerance_2.ok, relations].reset_index(drop=True)
itinerance_2 = itinerance_2.loc[:, relations]
print("\nnombre d'enregistrements sans erreurs : ", len(itinerance_ok_3))
print("nombre d'enregistrements avec au moins une erreur : ", len(itinerance_3))

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance            200
nom_enseigne - id_station_itinerance                 457
coordonneesXY - id_station_itinerance                624
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                  204
implantation_station - id_station_itinerance         136
nbre_pdc - id_station_itinerance                     513
condition_acces - id_station_itinerance                5
horaires - id_station_itinerance                     194
station_deux_roues - id_station_itinerance           521
adresse_station - coordonneesXY                      587

nombre d'enregistrements sans erreurs :  7649
nombre d'enregistrements avec au moins une erreur :  2166


-----------------------
## 6 - Cohérence date maj stations
- 20% des erreurs résiduelles sont liées au mélange d'anciens et de nouveaux pdc
- la suppression des anciens pdc permet de réduire de 25% le nombre de lignes erronnées ( 1 616 )
- les dernières erreurs correspondent à 468 stations associées à 32 opérateurs et sont liées à des causes multiples

In [338]:
itinerance_3['stat_maj'] = itinerance_3.id_station_itinerance + itinerance_3.date_maj
stat_maj_unique = itinerance_3.sort_values(by='stat_maj').drop_duplicates('id_station_itinerance', keep='last')
itinerance_3['last_stat_maj'] = itinerance_3['stat_maj'].isin(stat_maj_unique['stat_maj'])

doublons_stat_maj = itinerance_3[~itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)['index']
itinerance_4 = itinerance_3[itinerance_3['last_stat_maj']].loc[:, relations].reset_index(drop=True)
itinerance_3 = itinerance_3.loc[:, relations]
print('nombre de doublons stations : ', len(doublons_stat_maj))
print('nombre de pdc sans doublon   : ', len(itinerance_4))

nombre de doublons stations :  375
nombre de pdc sans doublon   :  1791


In [339]:
res = analyse_integrite(itinerance_4)
itinerance_ok_5 = itinerance_4.loc[itinerance_4.ok, relations].reset_index(drop=True)
itinerance_5_full = itinerance_4.loc[~itinerance_4.ok].reset_index(drop=True)
itinerance_5 = itinerance_5_full.loc[:, relations]
#itinerance_4 = itinerance_4.loc[:, relations]
print("\nnombre d'enregistrements sans erreurs : ", len(itinerance_ok_5))
print("nombre d'enregistrements avec au moins une erreur : ", len(itinerance_5))

index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance              0
nom_enseigne - id_station_itinerance                 267
coordonneesXY - id_station_itinerance                216
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                   40
implantation_station - id_station_itinerance          23
nbre_pdc - id_station_itinerance                     452
condition_acces - id_station_itinerance                0
horaires - id_station_itinerance                      53
station_deux_roues - id_station_itinerance           325
adresse_station - coordonneesXY                      414

nombre d'enregistrements sans erreurs :  175
nombre d'enregistrements avec au moins une erreur :  1616


In [340]:
il = Ilist.obj(itinerance_5)
print(il.tree())

-1: root-derived (1616)
   0 : index (1616)
      2 : nom_enseigne (317)
         9 : condition_acces (2)
      3 : coordonneesXY (413)
      4 : adresse_station (468)
         1 : contact_operateur (32)
      5 : id_station_itinerance (468)
         13: date_maj (82)
      6 : nom_station (432)
      7 : implantation_station (4)
      8 : nbre_pdc (21)
      10: horaires (49)
      11: station_deux_roues (5)
      12: id_pdc_itinerance (1616)
      14: last_modified (54)


In [324]:
# nb maxi d'erreurs
itinerance_4['somme'] = 0
for name in res.keys():
    itinerance_4['somme'] += 1 - itinerance_4[name]
erreurs = max(itinerance_4['somme'])
maxi = itinerance_4[itinerance_4.somme >= erreurs]
print("nombre d'enregistrements avec ", erreurs, " erreurs : ", len(maxi))

nombre d'enregistrements avec  4  erreurs :  12


-----------------------
## 7 - Synthèse


In [345]:
#itinerance_5_full.to_csv('erreurs_IRVE.csv')

itinerance['doublons_stat_maj'] = itinerance['index'].isin(doublons_stat_maj)
itinerance['doublons_pdc'] = itinerance['index'].isin(doublons_pdc)
itinerance['lignes_a_corriger'] = itinerance['index'].isin(itinerance_5['index'])
itinerance['doublons_a_supprimer'] = itinerance['doublons_stat_maj'] | itinerance['doublons_pdc']
itinerance['lignes_ko'] = itinerance['doublons_a_supprimer'] | itinerance['lignes_a_corriger']
print('total des lignes à corriger : ', itinerance['lignes_a_corriger'].sum())
print('total des doublons à supprimer : ', itinerance['doublons_a_supprimer'].sum())

total des lignes à corriger :  1616
total des doublons à supprimer :  8676


In [349]:
itinerance_ok = itinerance[~itinerance['lignes_ko']].loc[:, relations].reset_index(drop=True)
print('nombre de pdc : ', len(itinerance_ok))
print('\nbilan intégrité :')
res = analyse_integrite(itinerance_ok)

nombre de pdc :  40377

bilan intégrité :
index - id_pdc_itinerance                              0
contact_operateur - id_station_itinerance              0
nom_enseigne - id_station_itinerance                   0
coordonneesXY - id_station_itinerance                  0
id_station_itinerance - id_pdc_itinerance              0
nom_station - id_station_itinerance                    0
implantation_station - id_station_itinerance           0
nbre_pdc - id_station_itinerance                       0
condition_acces - id_station_itinerance                0
horaires - id_station_itinerance                       0
station_deux_roues - id_station_itinerance             0
adresse_station - coordonneesXY                        0


-----------------------
## 8 - Exemples d erreurs résiduelles 

### Erreurs multiples 
- exemple avec le nombre maximal d'erreurs (4) (2 stations soit 12 pdc avec : adresse, nom, nbre_pdc et deux-roues erronés)

In [310]:
maxi.loc[:, relations]

index              contact_operateur nom_enseigne         coordonneesXY  \
1536  47980  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1537  47981  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1538  47982  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1539  47983  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1540  47984  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1541  47985  assistance.technique@tevgo.fr       Belib'   [2.314663, 48.8737]   
1648  49154  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1649  49155  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1650  49156  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1651  49157  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1652  49158  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   
1653  49159  assistance.technique@tevgo.fr       Belib'  [2.360503, 48.86502]   

                    adresse_station id_station_itinerance  \
1536  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1537  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1538  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1539  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1540  37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1541   37 Rue la Boétie 75008 Paris          FRV75PPX0811   
1648     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1649     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1650     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1651     18 Rue Perrée 75003 Paris          FRV75PPX0304   
1652      18 Rue Perrée 75003 Paris          FRV75PPX0304   
1653     18 Rue Perrée 75003 Paris          FRV75PPX0304   

                    nom_station implantation_station  nbre_pdc  \
1536  Paris | Rue la Boétie 37               Voirie         5   
1537  Paris | Rue la Boétie 37               Voirie         5   
1538  Paris | Rue la Boétie 37               Voirie         5   
1539  Paris | Rue la Boétie 37               Voirie         5   
1540  Paris | Rue la Boétie 37               Voirie         5   
1541   Paris | Rue la Boétie 37               Voirie         6   
1648     Paris | Rue Perrée 18               Voirie         5   
1649     Paris | Rue Perrée 18               Voirie         5   
1650     Paris | Rue Perrée 18               Voirie         5   
1651     Paris | Rue Perrée 18               Voirie         5   
1652      Paris | Rue Perrée 18               Voirie         6   
1653     Paris | Rue Perrée 18               Voirie         5   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
1536     Accès libre     24/7               True     FRV75EPX08112   
1537     Accès libre     24/7               True     FRV75EPX08111   
1538     Accès libre     24/7              False     FRV75EPX08114   
1539     Accès libre     24/7              False     FRV75EPX08113   
1540     Accès libre     24/7              False     FRV75EPX08115   
1541     Accès libre     24/7              False     FRV75EPX08116   
1648     Accès libre     24/7              False     FRV75EPX03041   
1649     Accès libre     24/7              False     FRV75EPX03042   
1650     Accès libre     24/7              False     FRV75EPX03043   
1651     Accès libre     24/7              False     FRV75EPX03044   
1652     Accès libre     24/7               True     FRV75EPX03046   
1653     Accès libre     24/7               True     FRV75EPX03045   

        date_maj               last_modified  
1536  2022-12-16  2023-03-03T17:35:31.881000  
1537  2022-12-16  2023-03-03T17:35:31.881000  
1538  2022-12-16  2023-03-03T17:35:31.881000  
1539  2022-12-16  2023-03-03T17:35:31.881000  
1540  2022-12-16  2023-03-03T17:35:31.881000  
1541  2022-12-16  2023-03-03T17:35:31.881000  
1648  2022-12-16  2023-03-03T17:35:31.881000  
1649  2022-1

### Cohérence implantation_station - id_station
- 23 pdc sont liés à une erreur de choix d'implantation (3 stations, 1 opérateur)

In [302]:
itinerance_4.loc[~itinerance_4['implantation_station - id_station_itinerance'], relations]

index     contact_operateur nom_enseigne             coordonneesXY  \
1363  43183  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1364  43184  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1365  43185  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1366  43186  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1367  43187  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1368  43188  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1369  43190  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1372  43282  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1373  43283  contact@qovoltis.com     QOVOLTIS  [-2.7649063, 48.4941535]   
1375  43343  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1379  43579  contact@qovoltis.com     QOVOLTIS      [-0.69173, 44.83914]   
1380  43585  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1381  43586  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1382  43587  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1383  43588  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1384  43589  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1385  43590  contact@qovoltis.com     QOVOLTIS   [1.8088319, 43.6986226]   
1386  43595  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1387  43596  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1393  43611  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1394  43613  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1395  43614  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   
1397  43638  contact@qovoltis.com     QOVOLTIS      [-2.76491, 48.49415]   

                                        adresse_station id_station_itinerance  \
1363  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1364  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1365  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1366  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1367  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1368  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1369  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1372  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1373  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1375             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1379  19 AVENUE NEIL ARMSTRONG, Le Caucase 33700 MER...         FRQOVP3300015   
1380             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1381             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1382             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1383             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1384             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1385             1 AVENUE GEORGES POMPIDOU 81500 LAVAUR         FRQOVP8100001   
1386  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1387  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1393  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1394  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1395  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   
1397  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sain...         FRQOVP2200002   

                                        nom_station  \
1363  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1364  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1365  52 RUE DU GUÉ LAMBERT Hôtel Kyriad Direct Sai   
1366  19 AVENUE 

### Cohérence nom_station - id_station
- 40 pdc sont associés à une station avec un nom non cohérent

In [301]:
itinerance_4.loc[~itinerance_4['nom_station - id_station_itinerance'], relations]

index              contact_operateur              nom_enseigne  \
81     1752   service-recharge@soregies.fr               ALTERBASE86   
82     1753   service-recharge@soregies.fr               ALTERBASE86   
107    2862  support@lastmilesolutions.com       Last Mile Solutions   
108    2864  support@lastmilesolutions.com       Last Mile Solutions   
109    2877  support@lastmilesolutions.com       Last Mile Solutions   
110    2878  support@lastmilesolutions.com       Last Mile Solutions   
254    6412         support@alizecharge.fr  CPO Alizé Liberté Public   
256    6431         support@alizecharge.fr  CPO Alizé Liberté Public   
257    6435         support@alizecharge.fr  CPO Alizé Liberté Public   
310    7672         support@alizecharge.fr  CPO Alizé Liberté Public   
311    7693         support@alizecharge.fr  CPO Alizé Liberté Public   
822   22111    pascal.lhermitte@siege27.fr                   SIEGE87   
834   22125    pascal.lhermitte@siege27.fr                   SIEGE88   
861   22152    pascal.lhermitte@siege27.fr                   SIEGE49   
862   22153    pascal.lhermitte@siege27.fr                   SIEGE50   
1362  43174           contact@qovoltis.com                  QOVOLTIS   
1370  43214           contact@qovoltis.com                  QOVOLTIS   
1371  43216           contact@qovoltis.com                  QOVOLTIS   
1374  43288           contact@qovoltis.com                  QOVOLTIS   
1398  43639           contact@qovoltis.com                  QOVOLTIS   
1399  43640           contact@qovoltis.com                  QOVOLTIS   
1400  43643           contact@qovoltis.com                  QOVOLTIS   
1536  47980  assistance.technique@tevgo.fr                    Belib'   
1537  47981  assistance.technique@tevgo.fr                    Belib'   
1538  47982  assistance.technique@tevgo.fr                    Belib'   
1539  47983  assistance.technique@tevgo.fr                    Belib'   
1540  47984  assistance.technique@tevgo.fr                    Belib'   
1541  47985  assistance.technique@tevgo.fr                    Belib'   
1555  48380  assistance.technique@tevgo.fr                    Belib'   
1570  48414  assistance.technique@tevgo.fr                    Belib'   
1571  48415  assistance.technique@tevgo.fr                    Belib'   
1616  48506  assistance.technique@tevgo.fr                    Belib'   
1617  48507  assistance.technique@tevgo.fr                    Belib'   
1618  48509  assistance.technique@tevgo.fr                    Belib'   
1648  49154  assistance.technique@tevgo.fr                    Belib'   
1649  49155  assistance.technique@tevgo.fr                    Belib'   
1650  49156  assistance.technique@tevgo.fr                    Belib'   
1651  49157  assistance.technique@tevgo.fr                    Belib'   
1652  49158  assistance.technique@tevgo.fr                    Belib'   
1653  49159  assistance.technique@tevgo.fr                    Belib'   

                coordonneesXY  \
81        [0.37178, 46.56869]   
82        [0.37178, 46.56869]   
107     [5.585043, 47.149426]   
108     [5.585043, 47.149426]   
109     [5.585043, 47.149426]   
110     [5.585043, 47.149426]   
254     [1.45026, 43.6040701]   
256     [1.45026, 43.6040701]   
257     [1.45026, 43.6040701]   
310   [1.4503666, 43.6039735]   
311   [1.4503666, 43.6039735]   
822     [1.224898, 49.195263]   
834     [1.224898, 49.195263]   
861     [1.221452, 49.197647]   
862     [1.221452, 49.197647]   
1362  [4.0309288, 49.2849379]   
1370  [4.0309288, 49.2849379]   
1371  [4.0309288, 49.2849379]   
1374  [4.0309288, 49.2849379]   
1398    [4.0303445, 49.28518]   
1399    [4.0303445, 49.28518]   
1400    [4.0303445, 49.28518]   
1536      [2.314663, 48.8737]   
1537      [2.314663, 48.8737]   
1538      [2.314663, 48.8737]   
1539      [2.314663, 48.8737]   
1540      [2.314663, 48.8737]   
1541      [2.314663, 48.8737]   
1555   [2.3077698, 48.849297]   
1570   [2.3077698, 48.849297]   
1571   [2.3077698, 48.849297] 

### Cohérence adresse - coordonnées
- 414 pdc ont une adresse non cohérente avec les coordonnées géographiques

In [300]:
itinerance_4.loc[~itinerance_4['adresse_station - coordonneesXY'], relations]

index                        contact_operateur  \
0        10                   info@rossinienergy.com   
1        11                   info@rossinienergy.com   
2        12                   info@rossinienergy.com   
3        13                   info@rossinienergy.com   
4        17                   info@rossinienergy.com   
...     ...                                      ...   
1786  50861  supervision-ev.france@totalenergies.com   
1787  50862  supervision-ev.france@totalenergies.com   
1788  50863  supervision-ev.france@totalenergies.com   
1789  50864  supervision-ev.france@totalenergies.com   
1790  51263                   contact@ubitricity.com   

                   nom_enseigne                          coordonneesXY  \
0                 RossiniEnergy                  [3.262326, 49.856113]   
1                 RossiniEnergy                  [3.262326, 49.856113]   
2                 RossiniEnergy                  [3.262326, 49.856113]   
3                 RossiniEnergy                  [3.262326, 49.856113]   
4                 RossiniEnergy                  [3.262326, 49.856113]   
...                         ...                                    ...   
1786              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1787              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1788              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1789              RIESTER MEAUX  [2.87930851314442, 48.94679007929618]   
1790  Ubitricity France Network                  [0.193942, 49.544211]   

                              adresse_station id_station_itinerance  \
0     4 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
1     7 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
2     6 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
3     5 Avenue Archimède, 02100 Saint-Quentin             FRROSE352   
4     9 Avenue Archimède, 02100 Saint-Quentin             FRROSE351   
...                                       ...                   ...   
1786    52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1787    52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1788    51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1789    50 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1790                        Place Abb� Pierre        FRUBIE10058428   

                       nom_station          implantation_station  nbre_pdc  \
0                     Aventim_PC_1  Parking privé à usage public        10   
1                     Aventim_PC_1  Parking privé à usage public        10   
2                     Aventim_PC_1  Parking privé à usage public        10   
3                     Aventim_PC_1  Parking privé à usage public        10   
4                     Aventim_PC_2  Parking privé à usage public        10   
...                            ...                           ...       ...   
1786                 RIESTER MEAUX  Parking privé à usage public         4   
1787                 RIESTER MEAUX  Parking privé à usage public         4   
1788                 RIESTER MEAUX  Parking privé à usage public         4   
1789                 RIESTER MEAUX  Parking privé à usage public         4   
1790  Centre ville - Montivilliers                Parking public         1   

     condition_acces                                           horaires  \
0        Accès libre                                               24/7   
1        Accès libre                                               24/7   
2        Accès libre                                               24/7   
3        Accès libre                                               24/7   
4        Accès libre                                               24/7   
...              ...                                                ...   
1786     Accès libre  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...   
1787     Accès libre  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [298]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[0.193942, 49.544211]', relations]

index       contact_operateur               nom_enseigne  \
83     1866  contact@ubitricity.com  Ubitricity France Network   
1615  51263  contact@ubitricity.com  Ubitricity France Network   

              coordonneesXY    adresse_station id_station_itinerance  \
83    [0.193942, 49.544211]  Place Abbé Pierre        FRUBIE10056568   
1615  [0.193942, 49.544211]  Place Abb� Pierre        FRUBIE10058428   

                       nom_station implantation_station  nbre_pdc  \
83    Centre ville - Montivilliers       Parking public         1   
1615  Centre ville - Montivilliers       Parking public         1   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
83       Accès libre     24/7              FALSE    FRUBIE10056568   
1615     Accès libre     24/7              FALSE    FRUBIE10058428   

        date_maj               last_modified  
83    2023-01-18  2023-03-03T17:40:10.619000  
1615  2023-01-18  2023-03-03T16:22:10.908000

- exemple : une station avec plusieurs adresses -> erreur de saisie ?

In [297]:
itinerance_5.loc[itinerance_5.coordonneesXY	 == '[2.87930851314442, 48.94679007929618]', relations]

index                        contact_operateur   nom_enseigne  \
1611  50861  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1612  50862  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1613  50863  supervision-ev.france@totalenergies.com  RIESTER MEAUX   
1614  50864  supervision-ev.france@totalenergies.com  RIESTER MEAUX   

                              coordonneesXY  \
1611  [2.87930851314442, 48.94679007929618]   
1612  [2.87930851314442, 48.94679007929618]   
1613  [2.87930851314442, 48.94679007929618]   
1614  [2.87930851314442, 48.94679007929618]   

                            adresse_station id_station_itinerance  \
1611  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1612  52 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1613  51 Rue François De Tessan 77100 Meaux           FRTCBP00702   
1614  50 Rue François De Tessan 77100 Meaux           FRTCBP00702   

        nom_station          implantation_station  nbre_pdc condition_acces  \
1611  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1612  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1613  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   
1614  RIESTER MEAUX  Parking privé à usage public         4     Accès libre   

                                               horaires station_deux_roues  \
1611  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1612  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1613  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   
1614  Mo-Fr 08:30-12:00,Mo-Fr 13:30-19:00,Sa 09:00-1...              false   

     id_pdc_itinerance    date_maj               last_modified  
1611      FRTCBE003959  2023-01-27  2023-03-03T17:35:31.203000  
1612      FRTCBE003958  2023-01-27  2023-03-03T17:35:31.203000  
1613      FRTCBE003955  2023-01-27  2023-03-03T17:35:31.203000  
1614      FRTCBE003954  2023-01-27  2023-03-03T17:35:31.203000

### Cohérence station - coordonnées
- 216 pdc sont associés à des stations avec plusieurs adresses

In [299]:
itinerance_4.loc[~itinerance_4['coordonneesXY - id_station_itinerance'], relations]

index              contact_operateur              nom_enseigne  \
83     1769   service-recharge@soregies.fr               ALTERBASE86   
84     1782   service-recharge@soregies.fr               ALTERBASE86   
254    6412         support@alizecharge.fr  CPO Alizé Liberté Public   
256    6431         support@alizecharge.fr  CPO Alizé Liberté Public   
257    6435         support@alizecharge.fr  CPO Alizé Liberté Public   
...     ...                            ...                       ...   
1574  48430  assistance.technique@tevgo.fr                    Belib'   
1575  48431  assistance.technique@tevgo.fr                    Belib'   
1616  48506  assistance.technique@tevgo.fr                    Belib'   
1617  48507  assistance.technique@tevgo.fr                    Belib'   
1618  48509  assistance.technique@tevgo.fr                    Belib'   

                coordonneesXY  \
83        [0.84278, 46.67931]   
84      [0.842456, 46.679573]   
254     [1.45026, 43.6040701]   
256     [1.45026, 43.6040701]   
257     [1.45026, 43.6040701]   
...                       ...   
1574   [2.3036203, 48.869644]   
1575  [2.3036203, 48.8696448]   
1616   [2.3077698, 48.849297]   
1617  [2.3077697, 48.8492973]   
1618   [2.3077698, 48.849297]   

                                        adresse_station id_station_itinerance  \
83    Place de l'etoile - 86260 - Saint-Pierre de Ma...        FRS86PSOREB091   
84    Place de l'etoile - 86260 - Saint-Pierre de Ma...        FRS86PSOREB091   
254                  63 Bd Lazare Carnot 31000 Toulouse        FRTLSP31555040   
256                  63 Bd Lazare Carnot 31000 Toulouse        FRTLSP31555040   
257                  63 Bd Lazare Carnot 31000 Toulouse        FRTLSP31555040   
...                                                 ...                   ...   
1574                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1575                 248 Rue Pierre Charron 75008 Paris          FRV75PPX0801   
1616                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1617                    330 Avenue de Ségur 75007 Paris          FRV75PPX0704   
1618                   330 Avenue de Ségur 75007 Paris          FRV75PPX0704   

                                      nom_station implantation_station  \
83                         Saint-Pierre de Maillé               Voirie   
84                         Saint-Pierre de Maillé               Voirie   
254   TOULOUSE – Station Deux-Roues Lazare Carnot               Voirie   
256   TOULOUSE – Station Deux-Roues Lazare Carnot               Voirie   
257   TOULOUSE – Station Deux-Roues Lazare Carnot               Voirie   
...                                           ...                  ...   
1574               Paris | Rue Pierre Charron  62               Voirie   
1575               Paris | Rue Pierre Charron  62               Voirie   
1616                 Paris | Avenue de Ségur  55               Voirie   
1617                  Paris | Avenue de Ségur  55               Voirie   
1618                 Paris | Avenue de Ségur  55               Voirie   

      nbre_pdc condition_acces           horaires station_deux_roues  \
83           2     Accès libre               24/7              false   
84           2     Accès libre               24/7              false   
254          3     Accès libre  Mo-Su 00:00-23:59              FALSE   
256          3     Accès libre  Mo-Su 00:00-23:59              FALSE   
257          3     Accès libre  Mo-Su 00:00-23:59              FALSE   
...        ...             ...                ...                ...   
1574         3     Accès libre               24/7              False   
1575         4     Accès libre               24/7              False   
1616         5     Accès libre               24/7              False   
1617         6     Accès libre               24/7              False   
1618         5     Accès libre               24/7              False   

     id_pdc_itine

- exemple : une station avec plusieurs coordonnées -> incompréhension distinction station / pdc

In [292]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FR073PCAMAIEUFR', relations]

index         contact_operateur    nom_enseigne         coordonneesXY  \
1169  42830  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1170  42831  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   
1171  42832  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1172  42833  acelec@acelec-france.com  Camaïeu France  [3.207124,50.684984]   
1173  42834  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1174  42835  acelec@acelec-france.com  Camaïeu France  [3.207047,50.685013]   
1175  42836  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
1176  42837  acelec@acelec-france.com  Camaïeu France  [3.206962,50.685049]   
1179  42840  acelec@acelec-france.com  Camaïeu France   [3.207246,50.68494]   
1180  42841  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1181  42844  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1182  42845  acelec@acelec-france.com  Camaïeu France  [3.207433,50.684876]   
1183  42846  acelec@acelec-france.com  Camaïeu France  [3.207187,50.684964]   
1184  42847  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1185  42848  acelec@acelec-france.com  Camaïeu France   [3.20737,50.684898]   
1186  42849  acelec@acelec-france.com  Camaïeu France  [3.207306,50.684918]   

                         adresse_station id_station_itinerance  \
1169  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1170  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1171  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1172  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1173  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1174  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1175  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1176  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1179  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1180  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1181  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1182  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1183  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1184  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1185  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   
1186  211 Av. Jules Brame  59100 Roubaix       FR073PCAMAIEUFR   

         nom_station          implantation_station  nbre_pdc condition_acces  \
1169  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1170  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1171  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1172  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1173  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1174  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1175  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1176  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1179  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1180  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1181  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1182  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1183  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1184  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1185  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   
1186  CAMAÏEU FRANCE  Parking privé à usage public         1     Accès libre   

               horaires station_deux_roues id_pdc_itinerance    date_maj  \
1169  Mo-Su 08:00-08:00              false    FR073E8OV45243  2022-07-21   
1170  Mo-Su 08:00-08:00   

- exemple : une station avec plusieurs coordonnées -> erreur de saisie

In [289]:
itinerance_5.loc[itinerance_5.id_station_itinerance == 'FRV75PPX0704', relations]

index              contact_operateur nom_enseigne  \
1384  48380  assistance.technique@tevgo.fr       Belib'   
1399  48414  assistance.technique@tevgo.fr       Belib'   
1400  48415  assistance.technique@tevgo.fr       Belib'   
1445  48506  assistance.technique@tevgo.fr       Belib'   
1446  48507  assistance.technique@tevgo.fr       Belib'   
1447  48509  assistance.technique@tevgo.fr       Belib'   

                coordonneesXY                   adresse_station  \
1384   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1399   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1400   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1445   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   
1446  [2.3077697, 48.8492973]   330 Avenue de Ségur 75007 Paris   
1447   [2.3077698, 48.849297]  330 Avenue de Ségur 75007 Paris   

     id_station_itinerance                   nom_station implantation_station  \
1384          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1399          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1400          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1445          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   
1446          FRV75PPX0704   Paris | Avenue de Ségur  55               Voirie   
1447          FRV75PPX0704  Paris | Avenue de Ségur  55               Voirie   

      nbre_pdc condition_acces horaires station_deux_roues id_pdc_itinerance  \
1384         5     Accès libre     24/7              False     FRV75EPX07042   
1399         5     Accès libre     24/7              False     FRV75EPX07045   
1400         5     Accès libre     24/7              False     FRV75EPX07046   
1445         5     Accès libre     24/7              False     FRV75EPX07044   
1446         6     Accès libre     24/7              False     FRV75EPX07041   
1447         5     Accès libre     24/7              False     FRV75EPX07043   

        date_maj               last_modified  
1384  2022-12-16  2023-03-03T17:35:31.881000  
1399  2022-12-16  2023-03-03T17:35:31.881000  
1400  2022-12-16  2023-03-03T17:35:31.881000  
1445  2022-12-16  2023-03-03T17:35:31.881000  
1446  2022-12-16  2023-03-03T17:35:31.881000  
1447  2022-12-16  2023-03-03T17:35:31.881000

- exemple : une station par point de charge avec nbre_pdc = 1

In [254]:
itinerance_4.loc[itinerance_4.coordonneesXY == '[0.191899, 49.506711]', relations]

index       contact_operateur               nom_enseigne  \
7119  51257  contact@ubitricity.com  Ubitricity France Network   
7123  51261  contact@ubitricity.com  Ubitricity France Network   

              coordonneesXY                              adresse_station  \
7119  [0.191899, 49.506711]  Avenue Pr�sident Ren� Coty (Gare Beaulieu)    
7123  [0.191899, 49.506711]  Avenue Pr�sident Ren� Coty (Gare Beaulieu)    

     id_station_itinerance nom_station implantation_station  nbre_pdc  \
7119        FRUBIE10072897    Beaulieu               Voirie         1   
7123        FRUBIE10091696    Beaulieu               Voirie         1   

     condition_acces horaires station_deux_roues id_pdc_itinerance  \
7119     Accès libre     24/7              FALSE    FRUBIE10072897   
7123     Accès libre     24/7              FALSE    FRUBIE10091696   

        date_maj               last_modified  
7119  2023-01-18  2023-03-03T16:22:10.908000  
7123  2023-01-18  2023-03-03T16:22:10.908000

-----------------------
# Annexe : Initialisation du modèle de données 'mermaid'

In [1]:
from base64 import b64encode
from IPython.display import Image, display

def mermaid(graph):
  display(Image(url="https://mermaid.ink/img/" + b64encode(graph.encode("ascii")).decode("ascii")))

In [2]:
# modèle de données suivant l'application 'mermaid' 
irve = """ 
erDiagram
    AMENAGEUR ||..|{ STATION : amenage
    AMENAGEUR {
        string nom_amenageur
        string siren_amenageur
        string contact_amenageur 
    }
    OPERATEUR ||..|{ STATION : "exploite pour le compte de l enseigne"
    OPERATEUR {
        string contact_operateur PK "M"
        string nom_operateur 
        string telephone_operateur 
    }
    ENSEIGNE ||..|{ STATION : "heberge"
    ENSEIGNE {
        string nom_enseigne PK "M" 
    }
    STATION {
        string  id_station_itinerance PK "M"
        string  nom_station "M"
        enum    implantation_station "M"
        integer nbre_pdc "M"
        string  condition_acces "M"
        string  horaires "M"
        boolean station_deux_roues "M"
        string  id_station_local
        enum    raccordement
        string  num_pdl
        date    date_mise_en_service 
    }
    LOCALISATION ||--|{ STATION : "localise"
    LOCALISATION {
       array   coordonneesXY PK "M"
       string  adresse_station "M"
       string  code_insee_commune 
    }
    STATION ||--|{ POINT_DE_CHARGE : regroupe
    POINT_DE_CHARGE {
        string id_pdc_itinerance PK "M Root"
        number puissance_nominale "M"
        boolean prise_type_ef "M"
        boolean prise_type_2 "M"
        boolean prise_type_combo_ccs "M"
        boolean prise_type_chademo "M"
        boolean prise_type_autre "M"
        boolean paiement_acte "M"
        boolean paiement_autre "M"
        boolean reservation "M"
        enum    accessibilite_pmr "M"
        string  restriction_gabarit "M"
        date    date_maj "M"
        string  id_pdc_local
        boolean gratuit
        boolean paiement_cb
        string  tarification
        string  observations
        boolean cable_t2_attache 
    }
"""